File: Olympiad/Baltic/official/boi2018_solutions/code/dna/accepted/sl.py
#!/usr/bin/env python3
n,k,r = map(int, input().split())
v = list(map(int, input().split()))
counts = [0] * k
for x in v:
    counts[x] += 1

failed = 0
have = [0] * k
wanted = [0] * k
for i in range(r):
    sym, co = map(int, input().split())
    if counts[sym] < co:
        print("impossible")
        exit(0)
    wanted[sym] = co
    failed += 1

i = j = 0
res = 1000000000
while True:
    if failed > 0:
        if j == n:
            break
        have[v[j]] += 1
        if have[v[j]] == wanted[v[j]]:
            failed -= 1
        j += 1
    else:
        dif = j - i
        if dif < res:
            res = dif
        if have[v[i]] == wanted[v[i]]:
            failed += 1
        have[v[i]] -= 1
        i += 1
print(res)

File: Olympiad/Baltic/official/boi2018_solutions/code/dna/accepted/sl.cpp
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define trav(a, x) for(auto& a : x)
#define all(x) x.begin(), x.end()
#define sz(x) (int)(x).size()
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;

int main() {
	cin.sync_with_stdio(0); cin.tie(0);
	cin.exceptions(cin.failbit);
	int n, k, r;
	cin >> n >> k >> r;
	vi v(n);
	vi counts(k);
	rep(i,0,n) {
		cin >> v[i];
		assert(0 <= v[i] && v[i] < k);
		counts[v[i]]++;
	}
	int failed = 0;
	vi have(k), wanted(k);
	rep(i,0,r) {
		int sym, co;
		cin >> sym >> co;
		assert(0 <= sym && sym < k);
		assert(co > 0);
		if (counts[sym] < co) {
			cerr << "impossible" << endl;
			cout << "impossible" << endl;
			return 0;
		}
		wanted[sym] = co;
		++failed;
	}
	int i = 0, j = 0;
	int res = INT_MAX, occ = 0;
	for (;;) {
		if (failed) {
			if (j == n) break;
			if (++have[v[j]] == wanted[v[j]]) failed--;
			j++;
		} else {
			int dif = j - i;
			if (dif < res) res = dif, occ = 0;
			if (dif == res) occ++;
			assert(i < j);
			if (have[v[i]]-- == wanted[v[i]]) failed++;
			i++;
		}
	}
	assert(res != INT_MAX);
	cout << res << endl;
	cerr << res << ", " << occ << " places" << endl;
}
